### ISSUE TITLE
[BUG] File upload using POST via Send To doesn't work with CGI scripts

### ISSUE THREAD

**Issue by barnoid** [Waboodoo/HTTP-Shortcuts#186]
I'm trying to use HTTP-Shortcuts to allow me to send files from my phone to my web server using the Share to apps function. File upload using a POST request triggered by Send To fails when the receiving script uses the CGI libraries of either Python or Ruby, probably because of the lack of Content-Length header.

**Steps To Reproduce**
Set up a shortcut to POST to the CGI script on my server.
* Request Body / Parameters 
* * Request Body Type: Parameters (form-data)
* * One parameter called userfile type Single File
* Basic auth configured
* No other changes from defaults

In an app such as Photos or Files select a file and select the Share To icon, then select the Send to... icon from HTTP-Shortcuts.

On the server the following script can be called via Apache:
```
#!/usr/bin/env python3

import cgi

form = cgi.FieldStorage()

if 'userfile' in form and form['userfile'].file:
        with open('/path/filetestlog', 'w') as f:
                f.write(form['userfile'].filename)
                f.write(form['userfile'].type)
        with open('/path/filetest', 'wb') as f:
                f.write(form['userfile'].file.read())

        print("Content-Type: text/plain")
        print()
        print("OK")
else:   
        print("Content-Type: text/plain")
        print()
        print("NO FILE")
```
The files are writable by the web server.

**Expected behavior**
The file filetest contains the contents of the file sent from the app and filetestlog contains the file name and content type of the file on Android.
This is what happens if I call the same script using curl like so:
```
curl -v -X POST -F 'userfile=@testfile.jpg' https://server/cgi-bin/post_file.py
```

**Actual behavior**
The request works and the script returns OK but filetest is empty. filetestlog contains a file name in the form `shared_<uuid>.bin` and content type application/octet-stream regardless of file type sent.
With a similar CGI script in Ruby it fails with
```/usr/lib/ruby/2.5.0/cgi/core.rb:656:in `Integer': can't convert nil into Integer (TypeError)```
That line is attempting to parse a non-existent Content-Length header.

**Context:**
 - Android OS Version: 10
 - App Version 1.32.1
 - Server:
 - * Ubuntu 18.04
 - * Apache 2.4.29
 - * Python 3.6.9
 - * Ruby 2.5.1

**Comment by Waboodoo** [comment]
Thanks for the extensive bug report. I don't have the full setup but I can confirm that the app has so far not been sending the `Content-Type` header when using a file parameter or a file as the request body itself, so that is very likely the problem here. I'm fixing this issue for version 1.33.0, which will be available for [beta testing](https://play.google.com/apps/testing/ch.rmy.android.http_shortcuts) on the Play Store shortly.

**Comment by barnoid** [comment]
Thank you for the very useful app. Content-Length was the header that, from looking at the CGI library code, causes the problems I'm seeing when it's missing. Content-Type was always set to the same thing (application/octet-stream) as far as I could tell, but I could work around that.

I wrote another test script that just dumps the environment and stdin to a file without using any CGI library. That confirms that the complete file contents are being sent.

I'd be happy to beta test. Thanks again.

**Comment by barnoid** [comment]
It works! With 1.33.0. Thank you.

The file name and content-type are still set to generic values but these are less important problems. I could raise a separate bug for them?

**Comment by Waboodoo** [comment]
Android doesn't always provide the file name, so that one is mostly a guess / fallback. You can specify your own file name when setting up the parameter in the app, but it will be static (i.e., won't depend on the file you upload).

Similarly, the type of the file is also not always known, so again, it's mostly a guess. If the file type cannot be determine a fallback of "application/octet-stream" is used. I can look into this and see if this can maybe be improved a bit, but in general there is no optimal solution.

**Comment by barnoid** [comment]
Seems I spoke somewhat too soon. Ruby's CGI library doesn't like that the boundary lines in the request body don't have CR + LF line endings.

I think the problem lines are these two:

https://github.com/Waboodoo/HTTP-Shortcuts/blob/399472279101828a43f0887bd91b7611bb7c393b/HTTPShortcuts/app/src/main/kotlin/ch/rmy/android/http_shortcuts/http/FormMultipartRequestBody.kt#L46

https://github.com/Waboodoo/HTTP-Shortcuts/blob/399472279101828a43f0887bd91b7611bb7c393b/HTTPShortcuts/app/src/main/kotlin/ch/rmy/android/http_shortcuts/http/FormMultipartRequestBody.kt#L62

That's ok for the name and type. I can work around them. The file name would be more useful than the type if you do find a way.

Thanks again.

**Comment by Waboodoo** [comment]
I adapted the boundary lines to consistently use CR + LR. There is now a new version available for Beta testing in the Play Store which includes this fix.

**Comment by Waboodoo** [comment]
The fix for this is now released as part of version 1.33.0.
